src/java/org/apache/nutch/crawl/CrawlDbReader.java

```diff
@@ -185,13 +190,83 @@ public synchronized void write(Text key, CrawlDatum value)
         out.writeByte('\n');
       }
 
-      public synchronized void close(TaskAttemptContext context) throws IOException {
+      public synchronized void close(TaskAttemptContext context)
+          throws IOException {
+        out.close();
+      }
+    }
+
+    public RecordWriter<Text, CrawlDatum> getRecordWriter(
+        TaskAttemptContext context) throws IOException {
+      String name = getUniqueFile(context, "part", "");
+      Path dir = FileOutputFormat.getOutputPath(context);
+      FileSystem fs = dir.getFileSystem(context.getConfiguration());
+      DataOutputStream fileOut = fs.create(new Path(dir, name), context);
+      return new LineRecordWriter(fileOut);
+    }
+  }
+
+  public static class CrawlDatumJsonOutputFormat
+      extends FileOutputFormat<Text, CrawlDatum> {
+    protected static class LineRecordWriter
+        extends RecordWriter<Text, CrawlDatum> {
+      private DataOutputStream out;
+      private ArrayList<String> jsonString = new ArrayList<String>();
```

Comment:
Nutch is designed to be scalable and should be able to handle CrawlDbs containing billions of items. Even, if the CrawlDb is split into many partitions, every task has to handle 100 million CrawlDb items. However, it's hardly possible to keep such a huge list in memory, resp. it would be a too expensive use of resources. So, it's mandatory to write each record directly to the output, Hadoop itself handles the large outputs efficiently by buffering parts in memory and spilling to disk if the buffers are full.

To simplify the process: why not use the [JSON lines format](http://jsonlines.org/)? Every line one JSON-formatted record: `{"url": "...", ...}`
- makes it much easier to write the output because you need not to preserver state internally to properly open and close the list of records.
- JSON lines has several other advantages:
  - it's possible to grep for records
  - also [jq](https://stedolan.github.io/jq/) can process it
  - and again: any JSON parser needs to keep only a single line in memory